### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by FranziskaArndt**

### CODE REVIEW

File: src/ota_server.c [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+#include "esp_eth.h"
+#include "lwip/ip.h"
+#endif
+#include "send_eeprom.h"
+#include "ota_update.h"
+#include <errno.h>
+
+static const char *TAG = "ota-server";
+
+#define OTA_SERVER_PORT 1234
```
Comment (KlausPopp) [inline src/ota_server.c]: Make OTA_SERVER_PORT compile time configurable

File: examples/device/ethernet_ota/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# The following lines of boilerplate have to be in your project's CMakeLists
+# in this exact order for cmake to work correctly
+cmake_minimum_required(VERSION 3.5)
+
+set(EXCLUDE_COMPONENTS asio coap)
+set(EXTRA_COMPONENT_DIRS ../esp_ota_update_server)
```
Comment (KlausPopp) [inline examples/device/ethernet_ota/CMakeLists.txt:6]: Path is wrong. Must be
set(EXTRA_COMPONENT_DIRS ../../../../esp_ota_update_server)

File: examples/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+│   └── ethernet_ota      # update server example with the ethernet interface
+|       ├── main          # source code
+|       └── server_certs  # server certificate folder
+└── host
+    └── https_server      # https server running on the host
+```
+
+## Usage
+
+Copy the example of your choice next to the folder of the ota update server component. Build it with the esp-idf framework and flash the binary onto your ESP32(S2) board. Connect the microcontroller via USB with the host computer. For the IP-Address assignment check the README of the corresponding example.
```
Comment (KlausPopp) [inline examples/README.md]: Add example

File: examples/host/https_server/https_sever.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        httpd = http.server.HTTPServer((self.server_ip, self.server_port), http.server.SimpleHTTPRequestHandler)
+        httpd.socket = ssl.wrap_socket(httpd.socket,
+                                       keyfile=key_file,
+                                       certfile=server_file, server_side=True)
+        httpd.serve_forever()
+
+if __name__ == '__main__':
+    https_thread = HTTPS_Server('192.168.7.10', 8070)
+    https_thread.start()
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (KlausPopp) [inline examples/host/https_server/https_sever.py:31]: Add sleep. On RPI, server is sometimes not starting fast enough

File: src/ota_server.c [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+    /* start Ethernet driver state machine */
+    ESP_ERROR_CHECK(esp_eth_start(eth_handle));
+}
+
+#endif 
+
+
+void init_ota(void)
+{
+    check_current_partition();
```
Comment (KlausPopp) [inline src/ota_server.c]: Better call this later, after server has successfully started.

File: examples/device/cdc_ecm_ota/sdkconfig.defaults [hunk 6]
```diff
@@ -0,0 +1,8 @@
+CONFIG_IDF_TARGET="esp32s2"
+CONFIG_PARTITION_TABLE_TWO_OTA=y
+CONFIG_ESPTOOLPY_FLASHSIZE_4MB=y
+CONFIG_USB_ENABLED=y
+CONFIG_USB_NET_ENABLED=y
+CONFIG_LWIP_IPV6=n
+CONFIG_LWIP_NETIF_STATUS_CALLBACK=y
+CONFIG_FREERTOS_HZ=1000
```
Comment (KlausPopp) [inline examples/device/cdc_ecm_ota/sdkconfig.defaults:11]: Add CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE =y

File: src/ota_server.c [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+#if CONFIG_OTA_CDC_ECM
+/* configure cdc-ecm driver to use dhcp */
+static const ip_addr_t ipaddr  = IPADDR4_INIT_BYTES(0, 0, 0, 0);
+static const ip_addr_t netmask = IPADDR4_INIT_BYTES(0, 0, 0, 0);
+static const ip_addr_t gateway = IPADDR4_INIT_BYTES(0, 0, 0, 0);
+#endif
+
+void ota_server_task(void *pvParameters)
+{
+    char addr_str[128];
```
Comment (KlausPopp) [inline src/ota_server.c:44]: 1) The server task should be monitored by the watchdog
2) The watchdog timeout shall cause a reboot

**Review by KlausPopp — CHANGES_REQUESTED** [review]

File: Kconfig [hunk 8]
```diff
@@ -14,6 +14,12 @@ menu "OTA Update Server"
     
     endchoice # INTERFACE_TYPE
 
+    config OTA_SERVER_PORT
```
Comment (KlausPopp) [inline Kconfig:17]: Update the README ("This component provides a tcp server which awaits a message on port 1234..."

File: README.md [hunk 9]
```diff
@@ -8,6 +8,11 @@ This component provides a tcp server which awaits a message on port 1234 with a
 
 In the configuration, the EEPROM Read feature can be enabled and disabled. If the feature is enabled, the tcp server sends the content of a external EEPROM to the connected client directly after the connection was accepted and before the client shall send the URL. The EEPROM is connected via I2C with the ESP32(S2). The I2C connection can be configured in the configuration menu.
 
+## Task Watchdog
+The ota update server is monitored by a task watchdog, which has to be triggered within a certain time. The allowed time window can be adjusted via the setting CONFIG_ESP_TASK_WDT_TIMEOUT_S in the configuration menu. If the ota update server is in standby and waits for a connecting client, it triggers the watchdog every 3 seconds.
```
Comment (KlausPopp) [inline README.md:12]: What happens if the ota server crashes after the diagnostic cb has returned true? Then, the module is just restarted, but no roll-back. Correct?

File: README.md [hunk 10]
```diff
@@ -8,6 +8,11 @@ This component provides a tcp server which awaits a message on port 1234 with a
...
+
+ATTENTION: During the update the watchdog is not triggered. How long this may take depends on the size of the image. This time must be considered when configuring the watchdog timeout.
```
Comment (KlausPopp) [inline README.md:14]: The maximum image size is known (from the partition table). The default timeout shall be set to allow flashing of one OTA partition.

File: examples/device/cdc_ecm_ota/sdkconfig.defaults [hunk 11]
```diff
@@ -1,6 +1,9 @@
 CONFIG_IDF_TARGET="esp32s2"
+CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE =y
 CONFIG_PARTITION_TABLE_TWO_OTA=y
 CONFIG_ESPTOOLPY_FLASHSIZE_4MB=y
+CONFIG_ESP_TASK_WDT_PANIC=y
+CONFIG_ESP_TASK_WDT_TIMEOUT_S=20
```
Comment (KlausPopp) [inline examples/device/cdc_ecm_ota/sdkconfig.defaults:6]: The maximum image size is known (from the partition table). The default timeout shall be set to allow flashing of one OTA partition.

File: examples/device/ethernet_ota/sdkconfig.defaults [hunk 12]
```diff
@@ -1,6 +1,9 @@
 CONFIG_IDF_TARGET="esp32"
+CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE =y
 CONFIG_PARTITION_TABLE_TWO_OTA=y
 CONFIG_ESPTOOLPY_FLASHSIZE_4MB=y
+CONFIG_ESP_TASK_WDT_PANIC=y
+CONFIG_ESP_TASK_WDT_TIMEOUT_S=20
```
Comment (KlausPopp) [inline examples/device/ethernet_ota/sdkconfig.defaults:6]: The maximum image size is known (from the partition table). The default timeout shall be set to allow flashing of one OTA partition.

File: examples/device/ethernet_ota/sdkconfig.defaults [hunk 13]
```diff
@@ -1,6 +1,9 @@
 CONFIG_IDF_TARGET="esp32"
```
Comment (KlausPopp) [inline examples/device/ethernet_ota/sdkconfig.defaults:1]: CONFIG_FREERTOS_HZ=1000?

File: examples/host/https_server/https_sever.py [hunk 14]
```diff
@@ -26,6 +27,7 @@ def run(self):
 if __name__ == '__main__':
     https_thread = HTTPS_Server('192.168.7.10', 8070)
     https_thread.start()
+    time.sleep(2)
     s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
     s.connect(("192.168.7.1", 1234))
     data = s.recv(EEPROM_SIZE)
```
Comment (KlausPopp) [inline examples/host/https_server/https_sever.py:34]: Add comment: # remove this line if your ESP has no EEPROM attached

File: src/ota_server.c [hunk 15]
```diff
@@ -85,34 +91,58 @@ void ota_server_task(void *pvParameters)
...
+    /* register socket to wait to become ready */
+    memset(fds, 0 , sizeof(fds));
+    fds[0].fd = listen_sock;
+    fds[0].events = POLLIN;
+    /* set number of descriptors in fds */
+    nfds = 1;
+
     while (1) {
+        /* wait for listen_sock become ready to accept a connection */
+        int ret =  poll(fds, nfds, timeout);
```
Comment (KlausPopp) [inline src/ota_server.c:106]: Better compute nfds from size of fds

File: src/ota_server.c [hunk 16]
```diff
@@ -305,7 +335,7 @@ static void initialize_ethernet_interface(void)
 
     esp_eth_config_t config = ETH_DEFAULT_CONFIG(mac, phy);
     esp_eth_handle_t eth_handle = NULL;
-    ESP_ERROR_CHECK(esp_eth_driver_install(&config, &eth_handle));
+    ESP_ERROR_CHECK(esp_eth_driver_instmainall(&config, &eth_handle));
```
Comment (KlausPopp) [inline src/ota_server.c]: Is this change really intended?

**Review by KlausPopp — CHANGES_REQUESTED** [review]

File: examples/device/ethernet_ota/sdkconfig.defaults (hunk 13 above)
Comment (FranziskaArndt) [reply examples/device/ethernet_ota/sdkconfig.defaults:1]: Not necessary here, it is needed for the cdc-ecm interface to enable a small sleep in the driver. The ethernet interface do not use this driver.

File: src/ota_server.c (hunk 16 above)
Comment (FranziskaArndt) [reply src/ota_server.c]: No, that was a mistake

File: README.md (hunk 9 above)
Comment (FranziskaArndt) [reply README.md:12]: Yes, like every other critical error at this point.

File: examples/device/cdc_ecm_ota/sdkconfig.defaults (hunk 11 above)
Comment (FranziskaArndt) [reply examples/device/cdc_ecm_ota/sdkconfig.defaults:6]: 20 seconds are already the maximum timeout required

File: examples/device/ethernet_ota/sdkconfig.defaults (hunk 13 above)
Comment (KlausPopp) [reply examples/device/ethernet_ota/sdkconfig.defaults:1]: Ok for the example application this is not relevant. But when we create a base component for our firmware it should be configured identically for Ethernet and CDC-ECM. Because CONFIG_FREERTOS_HZ has an impact on the applciation behaviour.

File: examples/device/ethernet_ota/CMakeLists.txt (hunk 2 above)
Comment (KlausPopp) [reply examples/device/ethernet_ota/CMakeLists.txt:6]: Ok, path is correct, if example is copied as described in README

**Review by KlausPopp — APPROVED** [review]
